Update TemplateModel.php

BrainforgeUK opened the pull request:
Avoids unexpected message on template admin.

Pull Request for Issue #40666 .

### Summary of Changes

Improved handling of getMediaFiles() function.

### Testing Instructions

Install a child template without any media files.
[tpl_cassiopeia_test1.zip](https://github.com/joomla/joomla-cms/files/11581661/tpl_cassiopeia_test1.zip)

### Actual result BEFORE applying this Pull Request

Error message on template admin page
![unexpectedmessage](https://github.com/joomla/joomla-cms/assets/3941269/ff030eab-6d25-4408-ba40-0bb77ecf0fda)

### Expected result AFTER applying this Pull Request

![unexpectedmessage](https://github.com/joomla/joomla-cms/assets/3941269/a2f670be-a01f-4e9c-a9eb-349c3f5bc4b8)


### Link to documentations
Please select:
- [ ] Documentation link for docs.joomla.org: <link>
- [*] No documentation changes for docs.joomla.org needed

- [ ] Pull Request link for manual.joomla.org: <link>
- [*] No documentation changes for manual.joomla.org needed

BrainforgeUK committed:
Update TemplateModel.php

Avoids unexpected message on template admin.

ReLater commented:
1)
I'm not a friend of creating a bunch of empty folders if I explicitly don't need them or don't want them. Simply remove the tag `<inheritable>0</inheritable>` in your manifest.

References:
https://github.com/joomla/joomla-cms/pull/39329
But has been reverted: https://github.com/joomla/joomla-cms/pull/39943

2) Question:
If your solution is accepted. Will these folders automatically removed on uninstallation of the child template?

dgrammatiko commented:
> Install a child template without any media files.

This is not supported by Joomla, Joomla's own functionality WILL NEVER create a child **without** the expected folders. Instead of trying to bend Joomla to your way why not align your way to the Joomla's expectations?

BrainforgeUK commented:
1. Agree with empty folder comment.
Don't know why <inheritable> relevant.

2. As I mentioned in the related issue, this may not be relevant in 'the real world'!

I suppose when removing 'mytemplate', for completness, the installer should remove everything in the following places regardless of what the manifest XML says ... it may already do that, I have not investigated!

templates/mytemplate
media/templates/site/mytemplate
language/*/mytemplate.ini
language/*/mytemplate.sys.ini

There might be other areas to which a similar installer 'for completeness' comment might apply!

BrainforgeUK commented:
Then why does Joomla allow me to install a template without media files?

Surely the template media files will only be relevant to the template or other template related extensions. That's an issue for template / extension developers to address.

Joomla should just allow the situation without confusing messages during template testing / development.

dgrammatiko commented:
Also if you **must** do the magic creation either do it in the install or discover on the installer adapter: https://github.com/joomla/joomla-cms/blob/6e57d757c86e81fdb099995ce98a95c1af35e81d/libraries/src/Installer/Adapter/TemplateAdapter.php#L32

I'm not a fan of code that does something different than what is defined in the XML...

brianteeman commented:
> Joomla should just allow the situation without confusing messages during template testing / development.

absolutely not. its not a confusing message it is the 100% correct message.

BrainforgeUK commented:
I agree a change to the installer adapter would avoid this situation happening - and would need further investigation.
This change addresses the situation after it has happened in a dev / test environment.

This change provides a safe display of an empty media location to which the admin can then add folders / files in the usual manner.

A better way might be to create an empty index.html file in that location - but that does not solve the problem as index.html is not a media file! See functions getDirectoryTree() and checkFormat() ... would need an empty file called 'empty.jpg' or similar.

Just discovered that if admin deletes all items under media/templates/site/mytemplate there is no way they can add something. They have to add what they want and then delete what they don't want! My solution solves that problem, although they may not want the css, images, js and sccs folders I created ,,, they still have to create what they want and delete what I created! At least they can create them.

BrainforgeUK commented:
The message has not changed. The code change only avoids the message when it is unnecessary.

dgrammatiko commented:
> This change addresses the situation after it has happened in a dev / test environment.


Again: Joomla's own functionality produces child templates with the expected structure. To get to a point where the structure is not the expected you, probably, tried to manually create a child but missed the media folders. Joomla is not broken.

> This change provides a safe display of an empty media location to which the admin can then add folders / files in the usual manner.

Sorry this is kinda unacceptable (imho), the templates file manager is not supposed to correct faulty XML definitions, separation of concerns. 

Maybe https://github.com/dgrammatiko/mod_child_export will help your situation (?).The way you produce the installable child templates is quite crucial!

BrainforgeUK commented:
This is not a faulty XML definition.
It is a valid child template manifest.

dgrammatiko commented:
> It is a valid child template manifest.

Can you post the XML?

BTW, the `<media>` section is required, so are the media folders. ie: Joomla will never create a child without the media section in the XML or missing the folders...

BrainforgeUK commented:
Refer to pull request https://github.com/joomla/joomla-cms/pull/40665

dgrammatiko commented:
> Refer to pull request https://github.com/joomla/joomla-cms/pull/40665

You know what? Enough time wasted here, do whatever you want. Unsubscribing

richard67 commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1810,11 +1810,23 @@ public function getMediaFiles()
+
         if (!is_writable($path)) {
             $app->enqueueMessage(Text::_('COM_TEMPLATES_DIRECTORY_NOT_WRITABLE'), 'error');
         }
+        else if (empty($result)) {
+            // Add some default folders
+            foreach(['css', 'images', 'js', 'sccs'] as $folder)
+            {
```
```suggestion
        } elseif (empty($result)) {
            // Add some default folders
            foreach(['css', 'images', 'js', 'sccs'] as $folder) {
```
Fix PHP code style check failing in drone.

richard67 commented:
I agree with the above concerns that this PR is not the right way. But let's see if other opinions come it.

BrainforgeUK committed:
Update administrator/components/com_templates/src/Model/TemplateModel.php

Fixed PHP code style.

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

richard67 commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1810,11 +1810,21 @@ public function getMediaFiles()
-        }
+        } elseif (empty($result)) {
+            // Add some default folders
+            foreach(['css', 'images', 'js', 'sccs'] as $folder) {
```
```suggestion
            foreach (['css', 'images', 'js', 'sccs'] as $folder) {
```
Sorry, I've missed this last code style errors in my previous review.

BrainforgeUK committed:
Update administrator/components/com_templates/src/Model/TemplateModel.php

Style correction.

Co-authored-by: Richard Fath <richard67@users.noreply.github.com>

BrainforgeUK committed:
Update TemplateModel.php

Removed mkdir for css, image, js, sccs and replace entry result with current folder as that forces the empty folder to appear on the 'Templates: Customise' page, thus overcoming the original problem.

BrainforgeUK committed:
Update TemplateModel.php

Replaced tabs with spaces.

BrainforgeUK commented:
Better idea occurred to me avoiding the creation of dummy css, images, js and sccs folders. Appears to solve the problem and allows the site admin to add files / folders as needed.
![screenshot](https://github.com/joomla/joomla-cms/assets/3941269/866d0bb1-5db0-43f1-b003-2728835d390d)

brianteeman commented:
The only correct fix as stated before is to include the media folders.

BrainforgeUK commented:
That implies that the installer needs to be changed so that it either:
(a) It prevents attempts to install a template without a media section in the manifest. But what if there is no media content? Maybe unlikely but sounds to me like this places an unnecessary restriction on template developers. Anyway there is currently nothing to prevent a site admin using the 'Templates: Customise' page to manually remove all template media content.
or (b) If there is no media section in the manifest it creates the folder `media/templates/site/tpl_mytemplate` anyway.

In both cases the changes in this pull request serve 3 useful purposes which obviate the need for the above installer changes.
(1) On entry to the 'Templates: Customise' page ensures the `media/templates/site/tpl_mytemplate` folder exists anyway.
(2) If the `media/templates/site/tpl_mytemplate` folder is empty (missing in manifest or otherwise purged in some way) allows the site admin to use the 'Templates: Customise' page to see that it is empty and (re-)add folders and files as required.
(3) Only display the non-writtable warning if the `media/templates/site/tpl_mytemplate` folder is genuinely non-writtable. Currently the message is displayed if the folder does not exist, which is confusing - being non-writtable and missing are not the same thing!

wilsonge commented on administrator/components/com_templates/src/Model/TemplateModel.php:
```diff
@@ -1810,12 +1810,20 @@ public function getMediaFiles()
 
+        if (!is_dir($path)) {
+            // Just in case a template does not include any media
+            mkdir($path, 0777, true);
```
```suggestion
            Folder::create($path);
```

We shouldn't be creating any folders in core with 777

BrainforgeUK committed:
Update administrator/components/com_templates/src/Model/TemplateModel.php

Co-authored-by: George Wilson <georgejameswilson@googlemail.com>

HLeithner commented:
Thanks for your contribution. 
But as already explained, you have to provide everything Joomla needs on installation and add it to the manifest.

I'm closing this since it does unexpected magic from my point of view. You can always open it again if you think it's needed.
